tooling/bin/assess-minibroker-charts.rb

```diff
@@ -0,0 +1,622 @@
+  [
+    'mariadb',
+    'mongodb',
+    'postgresql',
+    'redis'
+  ]
+end
+
+def master_index
+  unless @master
+    puts "#{"Retrieving".cyan} master index ..."
+    @master = helm_index(stable)['entries']
+  end
+  @master
+end
+
+def helm_index(location)
+    uri = URI.parse (location + "/index.yaml")
+    res = Net::HTTP.get_response uri
+
+    # Debugging, save index data.
+    File.write(File.join(@workdir, 'index-location.txt'), uri)
+    File.write(File.join(@workdir, 'index.yaml'), res.body)
+
+    YAML.load (res.body)
+end
+
+def base_statistics
+  engines.each do |engine|
+    # Debugging. Save per-engine indices.
+    File.write(File.join(@workdir, "e-#{engine}.yaml"), master_index[engine].to_yaml)
+
+    # We are ignoring all the entries for which we do not have the
+    # engine version. Because that is the plan id later, therefore
+    # required.
+    puts "#{"Extracting".cyan} engine #{engine.blue}: #{master_index[engine].select do |chart|
+         chart ['appVersion']
+    end.length.to_s.cyan}"
+  end
+end
+
+def state
+  # Memoized
+  unless @state
+    # Look for state only in incremental mode. Do not fail if missing,
+    # just fall back to regular mode, starting with empty state.
+    if @incremental && File.exists?(statepath)
+      @state = YAML.load_file(statepath)
+    else
+      @state = {}
+    end
+    # State schema:
+    # <engine>.<version>.works	:: boolean
+    # <engine>.<version>.app	:: string
+    #
+    # version is chart version  (`version`)
+    # app     is engine version (`appVersion`)
+  end
+  @state
+end
+
+def state_save(engine, enginev, chartv, success)
+  @state[engine] = {} unless @state[engine]
+  @state[engine][chartv] = {
+    'app'   => enginev,
+    'works' => success,
+  }
+  File.write(statepath, @state.to_yaml)
+end
+
+def skip_chart(engine, enginev, chartv)
+  if @incremental && state[engine] && state[engine][chartv]
+    @skipped += 1
+    rewind
+    write "Skipping #{engine} #{enginev} #{chartv}"
+    # delay to actually see the output ?
+    true
+  else
+    false
+  end
+end
+
+def assess_chart (chart, engine, enginev, chartv, chartlocation)
+  log_start(engine, enginev, chartv)
+
+  rewind
+  write "  - #{engine.blue} #{enginev.blue}, chart #{chartv.blue} ..."
+
+  separator " helm repo setup ..." do
+    @the_repo = helm_repo_setup(chart, engine, chartlocation)
+
+    if @the_repo
+      " helm repo #{"up".green},"
+    else
+      state_save(engine, enginev, chartv, false)
+      " helm repo #{"start failed".repo}, likely a patch failure"
+    end
+  end
+
+  if @the_repo
+    separator " testing ..." do
+      success = do_test(@the_repo, engine)
+
+      state_save(engine, enginev, chartv, success)
+      if success
+        archive_save(engine, chartv)
+        regenerate_working_index
+        " testing #{"OK".green},"
+      else
+        " testing #{"FAIL".red},"
+      end
+    end
+
+    # clear leftovers, service & broker parts, ignoring errors.
```

Comment:
The cleanup (including `helm_repo_shutdown`) should be in a `rescue` block.

Reply:
Just `rescue` or `rescue some argument` ?

Reply:
No, like
```ruby
begin
  state_save_and_stuff
finally
  cf delete-service-offering-and-stuff
  helm_repo_shutdown
end
```
(Sorry, ended up actually meaning `finally` instead of `rescue`…)

Reply:
done